### PR TITLE
conf: change asserts to validations

### PR DESCRIPTION
**Pull request by IotaSpencer**
this is per https://wiki.python.org/moin/UsingAssertionsEffectively

### CODE REVIEW

File: conf.py [hunk 1]
```diff
@@ -16,6 +16,16 @@
 
 from . import world
 
+class ConfigValidationError(Exception):
+	"""Error when config conditions aren't met."""
+
```
Comment (GLolol) [inline conf.py]: Just this should be adequate (the base Exception class already handles arguments):

```python
class ConfigValidationError(Exception):
    """Error when config conditions aren't met."""
````

Also, please use 4 spaces to indent and not a tab! You can normally tell your editor to prefer this behaviour, at least for Python.

File: conf.py [hunk 2]
```diff
@@ -44,12 +54,19 @@
         }
 confname = 'unconfigured'
 
+def validate(condition, errmsg):
+	"""Function to wrap validation for use in validateConf"""
```
Comment (GLolol) [inline conf.py]: "wrap validation" is a bit vague: try something like `"""Convenience function to validate conditions in validateConf().""""`

File: conf.py [hunk 3]
```diff
@@ -44,12 +54,19 @@
...
+	if not condition:
+		raise ConfigValidationError(errmsg)
```
Comment (GLolol) [inline conf.py]: Ditto above, spaces and not tabs

File: conf.py [hunk 4]
```diff
@@ -44,12 +54,19 @@
...
+	
 def validateConf(conf, logger=None):
     """Validates a parsed configuration dict."""
-    assert type(conf) == dict, "Invalid configuration given: should be type dict, not %s." % type(conf).__name__
+    validate(type(conf) == dict,
+    "Invalid configuration given: should be type dict, not %s."
+    % type(conf).__name__)
```
Comment (GLolol) [inline conf.py]: Can you indent multiline code so it matches the the left parenthesis? Example:

```python
    validate(type(conf) == dict,
             "Invalid configuration given: should be type dict, not %s."
             % type(conf).__name__)
```

File: conf.py [hunk 5]
```diff
@@ -66,15 +83,17 @@ def validateConf(conf, logger=None):
...
+        validate(type(block.get('password')) == str, "Bad password %s for account %s" % (block.get('password'), account))
 
-    assert conf['login'].get('password') != "changeme", "You have not set the login details correctly!"
+    validate(conf['login'].get('password') != "changeme", "You have not set the login details correctly!")
 
     if newlogins:
-        assert conf.get('permissions'), "New-style accounts enabled but no permissions block was found. You will not be able to administrate your PyLink instance!"
+        validate(conf.get('permissions'),
+        "New-style accounts enabled but no permissions block was found."+\
+        "You will not be able to administrate your PyLink instance!")
```
Comment (GLolol) [inline conf.py]: Same as above (indentation)

**Comment by IotaSpencer** [comment]
**With login::user+password**:
Starts up how it should, no errors

**With login::accounts::... AND NO permissions**:
`ERROR: Failed to load config from 'pylink.yml': ConfigValidationError: New-style accounts enabled but no permissions block was found. You will not be able to administrate your PyLink instance!`

**With login::accounts::... AND permissions**: 
Starts up fine with no errors.

**Comment by IotaSpencer** [comment]
**Login removed**
`ERROR: Failed to load config from 'pylink.yml': ConfigValidationError: Missing 'login' section in config.`

**DeQuote password**:
yaml parses value as string either way

**Quote User**:
yaml parses key as string, $ gets parsed as string suprisingly, e.g. Io$ta and logs in correctly

**Login::Password == changeme**:
```
WARNING: The 'login:user' and 'login:password' options are deprecated since PyLink 1.1.
Please switch to the new 'login:accounts' format as outlined in the example config.
ERROR: Failed to load config from 'pylink.yml': ConfigValidationError:
You have not set the login details correctly!
```

Note: Error wrapping done manually so no scrollbars were present.

**Pull request merged**